proposals/0000-template.md

```diff
@@ -0,0 +1,154 @@
+
+# Motivation
+[motivation]: #motivation
+
+For a decentralized protocol, governance never fails to be one of the most important piece of the entire system. Without
+proper governance the network will likely fall apart. On-chain voting is a crucial constituent of governance and this NEP
+aims to make a first step in that direction by laying out the framework of a general-purpose voting contract that allows
+validators as a community to make decisions on chain. 
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The structure of this contract looks like the following:
+
+```rust
+pub type ProposalId = u64;
+
+#[near_bindgen]
+pub struct Poll {
+    /// Human readable description of the poll.
+    description: String,
+    /// All proposals for this poll.
+    proposals: Map<ProposalId, Proposal>,
+    /// Accounts that have participated in this poll and the corresponding stake voted.
+    accounts: Map<AccountId, Balance>,
+    /// Next proposal id.
+    next_proposal_id: ProposalId,
+    /// Threshold for closing the poll, i.e, if the ratio of stake on a certain proposal over total stake reaches
+    /// threshold, the poll is closed.
+    threshold: Fraction,
+    /// Voting result. `None` means the poll is still open.
+    result: Option<ProposalId>,
+}
+```
+
+On the top level we maintain a poll where there are multiple proposals that can be voted on. The poll keeps track of the
+proposals and accounts that have participated in the poll so far and their total voted stake. When the stake voted on a given
+proposal reaches the threshold for this poll, the proposal becomes final and the poll is closed.
+
+As an important constituent of the poll, `Proposal` contains all the voting information of this particular proposal,
+in addition to the actual content of the proposal:
+```rust
+pub struct Proposal {
+    /// Human readable description of the proposal.
+    description: String,
+    /// Serialized metadata of the proposal.
+    metadata: String,
+    /// When this proposal expires.
+    expiration_height: BlockHeight,
+    /// Current votes on this proposal.
+    votes: Map<AccountId, Balance>,    
+}
+```
+Here `metadata` is the json serialized content of the proposal. For example, if the poll is about when the network upgrade
+should happen, `metadata` will be something like `{"height": 10000}`. Validators can vote on multiple different proposals
+at the same time, provided that the sum of stake voted does not exceed their current stake.
+
+The voting contract has the following methods to allow creation of polls and proposals, as well as voting for proposals.
+
+```rust
+impl Poll {
+    /// Initialize the poll on some topic.
+    pub fn new(topic: String, threshold: Fraction) -> Self;
+    /// Create a proposal for a given poll.
+    pub fn create_proposal(&mut self, description: String, metadata: String);
```

Comment:
This is likely has to be payable function to avoid spamming the contract state. But it doesn't have to be.
Also it should return the `ProposalId` of the new proposal.